src/cache.rs

```diff
@@ -0,0 +1,341 @@
+
+    #[fail(display = "Not a file")]
+    NotAFile,
+
+    #[fail(display = "Failed to access filesystem")]
+    Io(#[fail(cause)] io::Error),
+}
+
+/// Utilities for a sym/cfi or object cache.
+#[derive(Debug, Clone)]
+pub struct Cache {
+    /// Cache identifier used for metric names.
+    pub name: &'static str,
+
+    /// Directory to use for storing cache items. Will be created if it does not exist.
+    pub cache_dir: Option<PathBuf>,
+
+    /// Time when this process started.
+    start_time: SystemTime,
+
+    /// Options intended to be user-configurable.
+    pub cache_config: CacheConfig,
+}
+
+impl Cache {
+    pub fn new<P: AsRef<Path>>(
+        name: &'static str,
+        cache_dir: Option<P>,
+        cache_config: CacheConfig,
+    ) -> Self {
+        Cache {
+            name,
+            cache_dir: cache_dir.map(|x| x.as_ref().to_owned()),
+            start_time: SystemTime::now(),
+            cache_config,
+        }
+    }
+
+    pub fn cleanup(&self) -> Result<(), CleanupError> {
+        log::info!("Cleaning up cache: {}", self.name);
+        let cache_dir = match self.cache_dir {
+            Some(ref x) => x,
+            None => return Err(CleanupError::NoCachingConfigured),
+        };
+
+        let entries = match catch_not_found(|| read_dir(cache_dir))? {
+            Some(x) => x,
+            None => {
+                log::warn!("Directory not found");
+                return Ok(());
+            }
+        };
+
+        for entry in entries {
+            let entry = entry?;
+            let path = entry.path();
+            if let Err(e) = self.try_cleanup_path(&path) {
+                log::error!("Failed to clean up {}: {}", path.display(), LogError(&e));
+                capture_fail(&e);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn try_cleanup_path(&self, path: &Path) -> Result<(), CleanupError> {
+        log::debug!("Looking at {}", path.display());
+        if path.is_file() {
+            if catch_not_found(|| self.check_expiry(path))?.is_none() {
+                log::info!("Removing {}", path.display());
+                catch_not_found(|| remove_file(path))?;
+            }
+
+            Ok(())
+        } else {
+            Err(CleanupError::NotAFile)
+        }
+    }
+
+    /// Validate cache expiration of path. If cache should not be used,
+    /// `Err(io::ErrorKind::NotFound)` is returned.  If cache is usable, `Ok(x)` is returned, where
+    /// `x` indicates whether the file should be touched before using.
+    fn check_expiry(&self, path: &Path) -> io::Result<bool> {
+        let metadata = path.metadata()?;
+
+        log::trace!("File length: {}", metadata.len());
+
+        // Immediately expire malformed items that have been created before this process started.
+        // See docstring of MALFORMED_MARKER
+        if MALFORMED_MARKER.len() as u64 == metadata.len() {
+            let created_at = metadata.created()?;
+            if created_at < self.start_time {
```

Comment:
This should check something similar to:

```rust
if created_at < cmp::max(self.start_time, now() - config.invalid_cache_timeout)
```

Reply:
Note that a run of `symbolicator cleanup` already unconditionally removes all invalid caches. This was intended behavior to achieve the same effect you want to have here (periodically clean invalid caches regardless of start time).

Reply:
Discussed IRL. @jan-auer wants to establish the principle that correctness should not rely on cleanup.